src/main/java/com/eerussianguy/blazemap/engine/client/BlazeMapClientEngine.java

```diff
@@ -159,23 +160,39 @@ public static void onChunkChanged(ChunkPos pos, String source) {
         activePipeline.onChunkChanged(pos);
     }
 
+    // This exists only as a temporary store of change events that fire before the client engine is ready to process them.
+    // This will likely be removed and replaced by a better system as part of BME-74
+    private static class ChangeEventStore {
+        public ResourceKey<Level> dimension;
+        public ChunkPos pos;
+        public List<MasterDatum> data;
+
+        ChangeEventStore(ResourceKey<Level> dimension, ChunkPos pos, List<MasterDatum> data) {
+            this.dimension = dimension;
+            this.pos = pos;
+            this.data = data;
+        }
+    }
+
     public static void submitChanges(ResourceKey<Level> dimension, ChunkPos pos, List<MasterDatum> data, String source) {
         isServerSource = true;
         mdSource = source;
 
-        // If the storage hasn't been set, that means the player login event hasn't been received yet
-        // and getPipeline will throw a NullPointerException that crashes the game.
-        // To avoid this, will wait until storage exists before submitting this MD
-        while (storage == null) {
-            try {
-                Thread.sleep(50); // Approx 1 tick wait
-            } catch (InterruptedException e) {
-                // This isn't supposed to happen, but I can't compile without it
-                e.printStackTrace();
+        // If the storage hasn't been set, that means the player login event hasn't been received yet and
+        // attempting to getPipeline will throw a NullPointerException that crashes the game.
+        // To avoid this, will store the changes until the storage has been initialised and then submit any
+        // previously saved events before submitting the current one.
+        // To be refactored in BME-74.
+        if (storage == null) {
+                BlazeMapClientEngine.changeEventQueue.add(new ChangeEventStore(dimension, pos, data));
+        } else {
+            while (changeEventQueue.peek() != null) {
+                ChangeEventStore changeEvent = changeEventQueue.remove();
+                getPipeline(changeEvent.dimension).insertMasterData(changeEvent.pos, changeEvent.data);
             }
```

Comment:
While I'd have done this differently, it works well enough that I'll let it be.

HOWEVER...

You shouldn't be dumping the queue here.
This should happen when the storage is assigned and pipelines can now be created, so that it is done only once rather than once plus a zero times loop every time new MD comes in, which is several times a second usually.

Reply:
Yeah, that was something I was going to defer to the BME-74 refactor. This should be all O(1) operations, so when the queue is empty it's just seeing if the item at the head of the queue is null, which should be a fairly trivial operation and therefore fairly fast. I mainly just didn't want to be doing too many big changes right now when the main goal is to unblock 1.20.1.

Reply:
I'll take this as a quick patch and then put it in place when inevitably the engine gets reworked in 1.20 🙂